### ISSUE TITLE
up- and downswiping in firefox under windows 8.1 touch devices needs 2 or more fingers

### ISSUE THREAD

**Issue by tripletto** [alvarotrigo/fullPage.js#1016]
I tried the swipe up, down, left and right gestures on two different win 8.1 devices with firefox 35.0.1 with your demo-page. Instead of moving through the sections, only the content gets selected.

When you use 2 or more fingers up- and downswipe work as "expected". Left- and rightswipe do not work at all.

This jquery-mobile demo seems to work as intended - with one finger.
http://api.jquerymobile.com/swipe/

**Comment by alvarotrigo** [comment]
When yo mean, touch devices you mean touch screen computers?

**Comment by tripletto** [comment]
yes. i could test that behavior on one tablet and one notebook both with windows 8.1.

**Comment by alvarotrigo** [comment]
I can not reproduce it by using Firefox on desktop with touch detection enabled in the developer tools. It scrolls horizontally and vertically. 

Have you used the [demo page](http://alvarotrigo.com/fullPage/) for your tests?

**Comment by tripletto** [comment]
yes, i did.

**Comment by alvarotrigo** [comment]
So, you are saying that you need two fingers to scroll to any direction but that in the case of the horizontal scrolling not even with that. Right?

**Comment by tripletto** [comment]
that is true. right now i tried the firefox responsive design view with touch detection enabled as you suggested: nothing at all happens as i click and drag on my windows 8.1 machine.

then i tried the same config in browserstack (windows 8.1 + firefox 35) with the same result - no swiping at all with the firefox responsive design view touch simulation.

![browserstack](https://cloud.githubusercontent.com/assets/10447009/6169958/41699c94-b2d1-11e4-9672-aa7c2d09a2dd.jpg)

you meant this tool from the screenshot, right? i tried that on browserstack with mac os yosemite and ff35, 34 and 30 - no success.

**Comment by alvarotrigo** [comment]
@tripletto what I'm saying is that **I can not reproduce the problem**. It works as expected to me. 
Make sure you reload the site after activating the touch input.

**Comment by alvarotrigo** [comment]
Tested in Yosemite, Windows 8.1 and Firefox 34 and 35.

**Comment by tripletto** [comment]
I don't get it. can someone with win 8.1 and touch please try?

are you using a touchpad with gesture support or a normal, plain mouse for testing? 

i just took my girlfriends laptop (win 8.1) (no touchscreen but normal touchpad) and with that, i am able to move the section with a expected two finger gesture in chrome and firefox but not in ie :D

**Comment by alvarotrigo** [comment]
@tripletto I'm talking about the Firefox developer tools. It works there as expected.

**Comment by tripletto** [comment]
I looked further into the issue here and found out, that I'm not batshit crazy :) and that the firefox touch simulator doesn't represent the firefox-reality at all!

Here is what I found: per default-config in windows, firefox doesn't fire touch-events. I tested with that: http://mobilehtml5.org/ts/?id=11

When I use 2 fingers with default-config firefox, I seems to emulate a normal touchpad! two fingers = mousewheel.

Touch events where deactivated by mozilla because of bugginess (text selection instead touch-events). You have to activate touch events in the about:config section but even then they are buggy and swiping won't work, which seems to be a firefox problem - not yours.

But this also means, that windows firefox users with touch devices have no touch-functionality with your script.

http://stackoverflow.com/questions/25024197/how-to-make-sure-touch-events-are-enabled-in-firefox-desk-top-browser

https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Touch_events

Best regards
tripletto

**Comment by alvarotrigo** [comment]
I see. Then the [swipe feature you were talking about](http://api.jquerymobile.com/swipe/) only works in Firefox because it works on mouse down, not on touch only. 

It seems there's no proper solution for it then.

**Comment by tripletto** [comment]
Maybe a pointer-events polyfill would work for desktop-browser "unification".
https://github.com/jquery/PEP
http://blog.jquery.com/2014/12/17/improving-the-pointer-events-polyfill/

**Comment by alvarotrigo** [comment]
> Touch events where deactivated by mozilla because of bugginess (text selection instead touch-events). You have to activate touch events in the about:config section but even then they are buggy and swiping won't work, which seems to be a firefox problem - not yours.
> 
> But this also means, that windows firefox users with touch devices have no touch-functionality with your script.

So... how do you usually scroll a webpage with firefox if the touch functionality doesn't work?
Does it works as a scroll event instead?

**Comment by tripletto** [comment]
yes, 1-finger-touch scrolls a normal (like github) page in firefox.

**Comment by alvarotrigo** [comment]
@tripletto I don't understand it very well... the mouse event still being binded in fullPage.js. It should work if its firing the "wheel" event.

**Comment by tripletto** [comment]
"Should" is the right expression. I have used Firebug's monitorEvents solution and the only thing this thing monitors are mousemove events. even if i use 2 fingers (which scrolls in fullpage.js and where i though that 2 fingers maybe emulate mousewheel)

https://getfirebug.com/wiki/index.php/MonitorEvents

I believe that firefox has its own way to deal with touch events. maybe the touch events never reach the DOM.

I made myself this fiddle (don't kill me :D I am more the designer type) to test the polyfill I talked about and no finger combination whatsoever triggers touchstart. everything in firefox (even 10 fingers) is mouse ;)

http://jsfiddle.net/0ej6o6ue/

http://jsfiddle.net/mmfxwux5/
^^this version is with PEP. all browsers pretty much behave the same ;)

https://github.com/jquery/PEP
http://blog.jquery.com/2014/12/17/improving-the-pointer-events-polyfill/

**Comment by alvarotrigo** [comment]
@tripletto if its using the mouse move it looks like its scrolling as if you where pressing the mouse wheel button on your mouse... I wonder if `mousedown` event also gets fired... because that'd be an easy way to solve the problem.

**Comment by tripletto** [comment]
i checked. mousedown gets fired ;)
![firebug](https://cloud.githubusercontent.com/assets/10447009/6216571/c2fb65a2-b60b-11e4-848e-56817683dfff.jpg)

**Comment by alvarotrigo** [comment]
Great then! It acts exactly as the 3rd button of the mouse then.
Will be solved for the next release! I was in fact working on the 3rd button feature to improve fullPage.js accessibility!

**Comment by tripletto** [comment]
awsome!

**Comment by alvarotrigo** [comment]
@tripletto let me know if it is fixed now in version 2.5.7!

**Comment by tripletto** [comment]
I assume you worked in this section?
```
		//binding the mousemove when the mouse's middle button is released
		container.mousedown(function(e){
			//middle button
			if (e.which == 2){
				oldPageY = e.pageY;
				container.on("mousemove", mouseMoveHandler);
			}
		});

		//unbinding the mousemove when the mouse's middle button is released
		container.mouseup(function(e){
			//middle button
			if (e.which == 2){
				container.off("mousemove");
			}
		});
```
When I change the mouse-button to "1" it works!

```
		//binding the mousemove when the mouse's middle button is released
		container.mousedown(function(e){
			//middle button
			if (e.which == 1){
				oldPageY = e.pageY;
				container.on("mousemove", mouseMoveHandler);
			}
		});

		//unbinding the mousemove when the mouse's middle button is released
		container.mouseup(function(e){
			//middle button
			if (e.which == 1){
				container.off("mousemove");
			}
		});
```

 I have swipe up and down functionality in firefox - yayyyy, great jon. I believe the number "2" you chose is the right mouse-button?!

One firefox-"problem" isn't solved: While you swipe the content gets selected but I find that negligible ;)

**Comment by alvarotrigo** [comment]
> When I change the mouse-button to "1" it works!

Nop. Be careful with that, that's the first click of the mouse. If you change it to 1 it will scroll when clicking in a desktop computer as well.

I though Firefox would use the middle button (number 2) which is the one used when you click down in the mouse wheel instead of scrolling it.
It doesn't make much sense to use the 1st click because as you say, it will select the text of the site.

Is this the normal behaviour with any other site in your touch screen when using Firefox? Does the text get selected when scrolling?

**Comment by tripletto** [comment]
yeah and also it breaks ie11 I just found out :D. nope that is no normal behavior. "normal" pages can be scrolled with 1-finger-touch without selecting any content. to select text you need to double-tab in firefox.

i usually say, i hate ie(6,7,8) but this time i make an exception for firefox...

**Comment by alvarotrigo** [comment]
@tripletto what I need then is you to show me which events get fired when swiping with **one single finger** in any other site.  (mousemove, mousedown, touchmove, touchstart...)

**Comment by tripletto** [comment]
I think you can't do a thing. I just checked github. When you touch the screen in firefox with one finger a mouseout gets fired and that's it. After that mouseout there is event silence. No touchstart, no touchmove. When you swipe sideways you get mousemoves but with classic up- and downswipe there is just silence.

I starting to believe, that mozilla has its own shitty control-layer on top of the DOM, which interprets certain gestures.

As long as mozilla doesn't get its act together with that nasty touch-bug, I'm beginning to think, there is nothing to be done anymore.

**Comment by alvarotrigo** [comment]
@tripletto so, tell me see if I understood it correctly:

- Firefox touch event detection is disabled by default
- Firefox on a touch screen computer can still scroll with a single finger swipe in any site with the default settings.

**Comment by tripletto** [comment]
true, true and true. and yes I tried on multiple devices: one tablet and one notebook.
https://wiki.mozilla.org/Gecko/Touch

**Comment by alvarotrigo** [comment]
As per the link you post it seems it should be working in Firefox OS, which... should be the one used in  a tablet, right?

About the other, not sure what's the difference between Desktop and Metro. But I really wonder how it scrolls without firing any event.

**Comment by tripletto** [comment]
As I said before: both devices are windows 8.1. Windows tablets have no own OS like iOS. On a windows tablet you can run all your favourite programms like photoshop, illustrator or even 3d studio max ;).

The metro-thingy means that there are two ui-modes in windows (totally stupid – everyone hates it). The metro-UI is the touch-only part of windows, which has its own set of applications. one of them was firefox-metro which was pulled early 2014 because nobody (1000 people tops) was using it.

**Comment by alvarotrigo** [comment]
@tripletto it just came to my mind. What you can do do is use `scrollBar:true`. This way firefox should detect the scrolling and snap to the section.

**Comment by chrisc10196** [comment]
I'm also having a similar issue. On a surface pro 3. Running Windows 8.1 pro. Touch swipe action works fine on Chrome but not on Firefox. Doesn't work on example page either (http://alvarotrigo.com/fullPage/#secondPage). Tried adjusting 'touchSensitivity' and  'normalScrollElementTouchThreshold'. Tried changing scrollbar to true. Running latest Mozilla in 'desktop mode'.

**Comment by alvarotrigo** [comment]
@tripletto could you try this for me please?
I have no way to access firefox in any touch screen device.

```javascript
var isFirefox = typeof InstallTrigger !== 'undefined'; // Firefox 1.0+		

//binding the mousemove when the mouse's middle button is released
container.mousedown(function (e) {
    //middle button
    if (e.which == 2) {
        oldPageY = e.pageY;
        container.on('mousemove', mouseMoveHandler);
    }

    //middle button
    if (touchDevice && isReallyTouch(e) && e.which == 1 && isFirefox) {
        oldPageY = e.pageY;
        container.on('mousemove', mouseMoveHandler);
    }
});

//unbinding the mousemove when the mouse's middle button is released
container.mouseup(function (e) {
    //middle button
    if (e.which == 2) {
        container.off('mousemove');
    }

    //middle button
    if (touchDevice && isReallyTouch(e) && e.which == 1 && isFirefox) {
        oldPageY = e.pageY;
        container.on('mousemove', mouseMoveHandler);
    }
});
```

I've created this fiddle as well with the portion of code: http://jsfiddle.net/3gv5eawd/

**Comment by kaiserhase** [comment]
So I figured this is the a bug similar to mine (technically reverse issue). In Edge browser on Windows Phone 10, using two fingers to pinch-zoom causes the script to continually scroll without stop, up then down then up ... Doesnt happen on your demo. 

Reason I recognized the issue is that I use Hotjar to record user activity on my site, and it recorded my visit and the bug. Albeit, it recorded my visit as from an android device...  Here is the link of the video: https://insights.hotjar.com/p?site=80897&recording=136157450&token=ec8b54d811ce87b25cc031d62f3e741b If you click "skip pauses" youll see it bounce as if i am repeatedly scrolling up and down, whereas I am only pinch-zooming with two fingers once or twice.

**Comment by alvarotrigo** [comment]
Can someone confirm if this issue takes place in the [Drag And Drop example](http://alvarotrigo.com/fullPage/extensions/dragAndMove.html) ?

**Comment by romanovma** [comment]
works in Drag And Drop example  - cannot see any issues

**Comment by alvarotrigo** [comment]
@romanovma thanks for the feedback!

**Comment by alvarotrigo** [comment]
Any others with touch screen devices that can verify it works well on the Drag And Drop example?
